exporter/googlecloudpubsubexporter/exporter.go

```diff
@@ -15,55 +15,203 @@
-	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/model/otlp"
 	"go.opentelemetry.io/collector/model/pdata"
 	"go.uber.org/zap"
+	"google.golang.org/api/option"
+	pubsubpb "google.golang.org/genproto/googleapis/pubsub/v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
 )
 
 const name = "googlecloudpubsub"
 
 type pubsubExporter struct {
-	instanceName string
-	logger       *zap.Logger
-
-	topicName string
-
-	//
-	userAgent string
-	ceSource  string
-	config    *Config
-	//
+	instanceName         string
+	logger               *zap.Logger
+	topicName            string
+	client               *pubsub.PublisherClient
+	cancel               context.CancelFunc
+	userAgent            string
+	ceSource             string
+	ceCompression        Compression
+	config               *Config
+	tracesMarshaler      pdata.TracesMarshaler
+	tracesWatermarkFunc  TracesWatermarkFunc
+	metricsMarshaler     pdata.MetricsMarshaler
+	metricsWatermarkFunc MetricsWatermarkFunc
+	logsMarshaler        pdata.LogsMarshaler
+	logsWatermarkFunc    LogsWatermarkFunc
 }
 
 func (*pubsubExporter) Name() string {
 	return name
 }
 
+type Encoding int
+
+const (
+	OtlpProtoTrace  Encoding = iota
+	OtlpProtoMetric          = iota
+	OtlpProtoLog             = iota
+)
+
+type Compression int
+
+const (
+	Uncompressed Compression = iota
+	GZip                     = iota
+)
+
+type WatermarkBehavior int
+
+const (
+	Current  WatermarkBehavior = iota
+	Earliest                   = iota
+)
+
 func (ex *pubsubExporter) start(ctx context.Context, _ component.Host) error {
+	ctx, ex.cancel = context.WithCancel(ctx)
+
+	if ex.client == nil {
+		copts := ex.generateClientOptions()
+		client, err := pubsub.NewPublisherClient(ctx, copts...)
+		if err != nil {
+			return fmt.Errorf("failed creating the gRPC client to Pubsub: %w", err)
+		}
+
+		ex.client = client
+	}
+	ex.tracesMarshaler = otlp.NewProtobufTracesMarshaler()
+	ex.tracesWatermarkFunc = currentTracesWatermark
+	ex.metricsMarshaler = otlp.NewProtobufMetricsMarshaler()
+	ex.metricsWatermarkFunc = currentMetricsWatermark
+	ex.logsMarshaler = otlp.NewProtobufLogsMarshaler()
+	ex.logsWatermarkFunc = currentLogsWatermark
 	return nil
 }
 
 func (ex *pubsubExporter) shutdown(context.Context) error {
+	if ex.client != nil {
+		ex.client.Close()
+		ex.client = nil
+	}
 	return nil
 }
 
-func (ex *pubsubExporter) Capabilities() consumer.Capabilities {
-	return consumer.Capabilities{
-		MutatesData: false,
+func (ex *pubsubExporter) generateClientOptions() (copts []option.ClientOption) {
+	if ex.userAgent != "" {
+		copts = append(copts, option.WithUserAgent(ex.userAgent))
 	}
+	if ex.config.Endpoint != "" {
+		if ex.config.Insecure {
+			var dialOpts []grpc.DialOption
+			if ex.userAgent != "" {
+				dialOpts = append(dialOpts, grpc.WithUserAgent(ex.userAgent))
+			}
+			conn, _ := grpc.Dial(ex.config.Endpoint, append(dialOpts, grpc.WithTransportCredentials(insecure.NewCredentials()))...)
```

Comment:
Is this error able to be captured elsewhere? 

Just worried it could have a dead conn or partially set connection and it "working" without actually doing work.

Reply:
> Is this error able to be captured elsewhere?

I think this is the only thing I didn't fix because this is only for testing/mocking.